packages/gitbook/src/components/TableOfContents/useScrollToActiveTOCItem.tsx

```diff
@@ -0,0 +1,55 @@
+'use client';
+
+import React from 'react';
+
+const TOCScrollContainerContext = React.createContext<string | null>(null);
+
+const SCROLL_OFFSET = 200;
+
+/**
+ * Provider that supplies the ID of the scroll container for the table of contents.
+ */
+export function TOCScrollContainerProvider(
+    props: React.PropsWithChildren<{ scrollContainerId: string }>,
+) {
+    const { children, scrollContainerId } = props;
+
+    return (
+        <TOCScrollContainerContext.Provider value={scrollContainerId}>
+            {children}
+        </TOCScrollContainerContext.Provider>
+    );
+}
+
+/**
+ * Scrolls the table of contents container to the page item when it becomes active,
+ * but only if the item is outside the viewable area of the container.
+ */
+export function useScrollToActiveTOCItem(tocItem: {
+    isActive: boolean;
+    linkRef: React.RefObject<HTMLAnchorElement>;
+}) {
+    const { isActive, linkRef } = tocItem;
+    const scrollContainerId = React.useContext(TOCScrollContainerContext);
+
+    React.useEffect(() => {
+        if (isActive && linkRef.current && scrollContainerId) {
+            const container = document.getElementById(scrollContainerId);
+
+            if (container) {
+                const itemTop = linkRef.current.offsetTop;
+                const itemBottom = itemTop + linkRef.current.offsetHeight;
+                const containerTop = container.scrollTop;
+                const containerBottom = containerTop + container.clientHeight;
+
+                // Only scroll if the TOC item is outside the viewable area of the container
+                if (itemTop < containerTop || itemBottom > containerBottom) {
+                    container.scrollTo({
+                        top: itemTop - SCROLL_OFFSET,
+                        behavior: 'smooth',
+                    });
+                }
+            }
```

Comment:
Have you tried using `scrollIntoView` / `scrollIntoViewIfNeeded` https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView ?

Reply:
When I think about it, I had a similar thing in the app last week. And you might not need to know the parent scroll div as `scrollIntoView` will automaticaly detect it, so maybe the link ref is enough (to be tested)

Reply:
Yeah I tried `scrollIntoView` using just the `linkRef` but that would not work for some reason. I think that because in that case the scrolling container is not the immediate ancestor. That's why I ended up with the solution of getting the element of the actual scroll container and trigger the scroll to the specific position.

But I'll give it another try maybe missing something.

Reply:
Ok not sure what I was doing wrong but works when I retried 😅. Made the change. Thanks

Reply:
@SamyPesse Ok so reverted back to using scrollTo in the end as I was hitting when using scrollIntoView specifically that depending on how far down the TOC item was it could potentially also scroll the page content too. I think this has to do with the behavior described [in this issue](https://github.com/w3c/csswg-drafts/issues/9452) which is that scrollIntoView will run the scroll against each ancestor element and because in our case page content and TOC are sharing one.

Let me know what you think. I would be happier to use the simpler scrollIntoView but couldn't find a way to make it work without hitting the issue above.